File: fraiseql-scala/src/main/scala/com/fraiseql/security/Security.scala
package com.fraiseql.security

// Role matching strategies
sealed trait RoleMatchStrategy {
  def value: String
}
object RoleMatchStrategy {
  case object Any extends RoleMatchStrategy {
    def value = "any"
  }
  case object All extends RoleMatchStrategy {
    def value = "all"
  }
  case object Exactly extends RoleMatchStrategy {
    def value = "exactly"
  }

  def fromString(s: String): Option[RoleMatchStrategy] = s.toLowerCase match {
    case "any"     => Some(Any)
    case "all"     => Some(All)
    case "exactly" => Some(Exactly)
    case _         => None
  }
}

// Authorization policy types
sealed trait AuthzPolicyType {
  def value: String
}
object AuthzPolicyType {
  case object Rbac extends AuthzPolicyType {
    def value = "rbac"
  }
  case object Abac extends AuthzPolicyType {
    def value = "abac"
  }
  case object Custom extends AuthzPolicyType {
    def value = "custom"
  }
  case object Hybrid extends AuthzPolicyType {
    def value = "hybrid"
  }

  def fromString(s: String): Option[AuthzPolicyType] = s.toLowerCase match {
    case "rbac"   => Some(Rbac)
    case "abac"   => Some(Abac)
    case "custom" => Some(Custom)
    case "hybrid" => Some(Hybrid)
    case _        => None
  }
}

// Custom authorization configuration
case class AuthorizeConfig(
    rule: String = "",
    policy: String = "",
    description: String = "",
    errorMessage: String = "",
    recursive: Boolean = false,
    operations: String = "",
    cacheable: Boolean = true,
    cacheDurationSeconds: Int = 300
) {
  def toMap: Map[String, Any] = Map(
    "rule" -> rule,
    "policy" -> policy,
    "description" -> description,
    "errorMessage" -> errorMessage,
    "recursive" -> recursive,
    "operations" -> operations,
    "cacheable" -> cacheable,
    "cacheDurationSeconds" -> cacheDurationSeconds
  )
}

// Role-based access control configuration
case class RoleRequiredConfig(
    roles: List[String] = List(),
    strategy: RoleMatchStrategy = RoleMatchStrategy.Any,
    hierarchy: Boolean = false,
    description: String = "",
    errorMessage: String = "",
    operations: String = "",
    inherit: Boolean = false,
    cacheable: Boolean = true,
    cacheDurationSeconds: Int = 300
) {
  def toMap: Map[String, Any] = Map(
    "roles" -> roles,
    "strategy" -> strategy.value,
    "hierarchy" -> hierarchy,
    "description" -> description,
    "errorMessage" -> errorMessage,
    "operations" -> operations,
    "inherit" -> inherit,
    "cacheable" -> cacheable,
    "cacheDurationSeconds" -> cacheDurationSeconds
  )
}

// Authorization policy configuration
case class AuthzPolicyConfig(
    name: String,
    policyType: AuthzPolicyType = AuthzPolicyType.Custom,
    description: String = "",
    rule: String = "",
    attributes: List[String] = List(),
    cacheable: Boolean = true,
    cacheDurationSeconds: Int = 300,
    recursive: Boolean = false,
    operations: String = "",
    auditLogging: Boolean = false,
    errorMessage: String = ""
) {
  def toMap: Map[String, Any] = Map(
    "name" -> name,
    "type" -> policyType.value,
    "description" -> description,
    "rule" -> rule,
    "attributes" -> attributes,
    "cacheable" -> cacheable,
    "cacheDurationSeconds" -> cacheDurationSeconds,
    "recursive" -> recursive,
    "operations" -> operations,
    "auditLogging" -> auditLogging,
    "errorMessage" -> errorMessage
  )
}

// Builder for custom authorization
class AuthorizeBuilder {
  private var rule = ""
  private var policy = ""
  private var description = ""
  private var errorMessage = ""
  private var recursive = false
  private var operations = ""
  private var cacheable = true
  private var cacheDurationSeconds = 300

  def withRule(r: String): AuthorizeBuilder = {
    rule = r
    this
  }
  def withPolicy(p: String): AuthorizeBuilder = {
    policy = p
    this
  }
  def withDescription(d: String): AuthorizeBuilder = {
    description = d
    this
  }
  def withErrorMessage(e: String): AuthorizeBuilder = {
    errorMessage = e
    this
  }
  def withRecursive(r: Boolean): AuthorizeBuilder = {
    recursive = r
    this
  }
  def withOperations(o: String): AuthorizeBuilder = {
    operations = o
    this
  }
  def withCacheable(c: Boolean): AuthorizeBuilder = {
    cacheable = c
    this
  }
  def withCacheDurationSeconds(d: Int): AuthorizeBuilder = {
    cacheDurationSeconds = d
    this
  }
  def build(): AuthorizeConfig = AuthorizeConfig(
    rule, policy, description, errorMessage, recursive, operations, cacheable, cacheDurationSeconds
  )
}

// Builder for RBAC
class RoleRequiredBuilder {
  private var roles = List[String]()
  private var strategy = RoleMatchStrategy.Any
  private var hierarchy = false
  private var description = ""
  private var errorMessage = ""
  private var operations = ""
  private var inherit = false
  private var cacheable = true
  private var cacheDurationSeconds = 300

  def withRoles(r: List[String]): RoleRequiredBuilder = {
    roles = r
    this
  }
  def withStrategy(s: RoleMatchStrategy): RoleRequiredBuilder = {
    strategy = s
    this
  }
  def withHierarchy(h: Boolean): RoleRequiredBuilder = {
    hierarchy = h
    this
  }
  def withDescription(d: String): RoleRequiredBuilder = {
    description = d
    this
  }
  def withErrorMessage(e: String): RoleRequiredBuilder = {
    errorMessage = e
    this
  }
  def withOperations(o: String): RoleRequiredBuilder = {
    operations = o
    this
  }
  def withInherit(i: Boolean): RoleRequiredBuilder = {
    inherit = i
    this
  }
  def withCacheable(c: Boolean): RoleRequiredBuilder = {
    cacheable = c
    this
  }
  def withCacheDurationSeconds(d: Int): RoleRequiredBuilder = {
    cacheDurationSeconds = d
    this
  }
  def build(): RoleRequiredConfig = RoleRequiredConfig(
    roles, strategy, hierarchy, description, errorMessage, operations, inherit, cacheable,
    cacheDurationSeconds
  )
}

// Builder for authorization policies
class AuthzPolicyBuilder(val name: String) {
  private var policyType = AuthzPolicyType.Custom
  private var description = ""
  private var rule = ""
  private var attributes = List[String]()
  private var cacheable = true
  private var cacheDurationSeconds = 300
  private var recursive = false
  private var operations = ""
  private var auditLogging = false
  private var errorMessage = ""

  def withType(t: AuthzPolicyType): AuthzPolicyBuilder = {
    policyType = t
    this
  }
  def withDescription(d: String): AuthzPolicyBuilder = {
    description = d
    this
  }
  def withRule(r: String): AuthzPolicyBuilder = {
    rule = r
    this
  }
  def withAttributes(a: List[String]): AuthzPolicyBuilder = {
    attributes = a
    this
  }
  def withCacheable(c: Boolean): AuthzPolicyBuilder = {
    cacheable = c
    this
  }
  def withCacheDurationSeconds(d: Int): AuthzPolicyBuilder = {
    cacheDurationSeconds = d
    this
  }
  def withRecursive(r: Boolean): AuthzPolicyBuilder = {
    recursive = r
    this
  }
  def withOperations(o: String): AuthzPolicyBuilder = {
    operations = o
    this
  }
  def withAuditLogging(a: Boolean): AuthzPolicyBuilder = {
    auditLogging = a
    this
  }
  def withErrorMessage(e: String): AuthzPolicyBuilder = {
    errorMessage = e
    this
  }
  def build(): AuthzPolicyConfig = AuthzPolicyConfig(
    name, policyType, description, rule, attributes, cacheable, cacheDurationSeconds,
    recursive, operations, auditLogging, errorMessage
  )
}

object Authorization {
  def authorize() = new AuthorizeBuilder()
  def roleRequired() = new RoleRequiredBuilder()
  def policy(name: String) = new AuthzPolicyBuilder(name)
}

File: fraiseql-csharp/src/FraiseQL.Security/Security.cs
using System;
using System.Collections.Generic;
using System.Linq;

namespace FraiseQL.Security
{
    /// <summary>
    /// Role matching strategies for RBAC
    /// </summary>
    public abstract record RoleMatchStrategy
    {
        public string Value { get; }

        private RoleMatchStrategy(string value)
        {
            Value = value;
        }

        /// <summary>At least one role must match</summary>
        public sealed record Any : RoleMatchStrategy
        {
            public Any() : base("any") { }
        }

        /// <summary>All roles must match</summary>
        public sealed record All : RoleMatchStrategy
        {
            public All() : base("all") { }
        }

        /// <summary>Exactly these roles</summary>
        public sealed record Exactly : RoleMatchStrategy
        {
            public Exactly() : base("exactly") { }
        }

        public static RoleMatchStrategy FromString(string value) =>
            value.ToLowerInvariant() switch
            {
                "any" => new Any(),
                "all" => new All(),
                "exactly" => new Exactly(),
                _ => throw new ArgumentException($"Unknown strategy: {value}")
            };
    }

    /// <summary>
    /// Authorization policy types
    /// </summary>
    public abstract record AuthzPolicyType
    {
        public string Value { get; }

        private AuthzPolicyType(string value)
        {
            Value = value;
        }

        /// <summary>Role-based access control</summary>
        public sealed record Rbac : AuthzPolicyType
        {
            public Rbac() : base("rbac") { }
        }

        /// <summary>Attribute-based access control</summary>
        public sealed record Abac : AuthzPolicyType
        {
            public Abac() : base("abac") { }
        }

        /// <summary>Custom authorization rules</summary>
        public sealed record Custom : AuthzPolicyType
        {
            public Custom() : base("custom") { }
        }

        /// <summary>Hybrid approach combining RBAC and ABAC</summary>
        public sealed record Hybrid : AuthzPolicyType
        {
            public Hybrid() : base("hybrid") { }
        }

        public static AuthzPolicyType FromString(string value) =>
            value.ToLowerInvariant() switch
            {
                "rbac" => new Rbac(),
                "abac" => new Abac(),
                "custom" => new Custom(),
                "hybrid" => new Hybrid(),
                _ => throw new ArgumentException($"Unknown policy type: {value}")
            };
    }

    /// <summary>
    /// Configuration for custom authorization rules
    /// </summary>
    public record AuthorizeConfig(
        string Rule = "",
        string Policy = "",
        string Description = "",
        string ErrorMessage = "",
        bool Recursive = false,
        string Operations = "",
        bool Cacheable = true,
        int CacheDurationSeconds = 300
    )
    {
        public Dictionary<string, object> ToDict() => new()
        {
            { "rule", Rule },
            { "policy", Policy },
            { "description", Description },
            { "errorMessage", ErrorMessage },
            { "recursive", Recursive },
            { "operations", Operations },
            { "cacheable", Cacheable },
            { "cacheDurationSeconds", CacheDurationSeconds }
        };
    }

    /// <summary>
    /// Configuration for role-based access control
    /// </summary>
    public record RoleRequiredConfig(
        List<string> Roles = null!,
        RoleMatchStrategy Strategy = null!,
        bool Hierarchy = false,
        string Description = "",
        string ErrorMessage = "",
        string Operations = "",
        bool Inherit = false,
        bool Cacheable = true,
        int CacheDurationSeconds = 300
    )
    {
        public RoleRequiredConfig() : this(
            new List<string>(),
            new RoleMatchStrategy.Any(),
            false,
            "",
            "",
            "",
            false,
            true,
            300
        ) { }

        public Dictionary<string, object> ToDict() => new()
        {
            { "roles", Roles },
            { "strategy", Strategy.Value },
            { "hierarchy", Hierarchy },
            { "description", Description },
            { "errorMessage", ErrorMessage },
            { "operations", Operations },
            { "inherit", Inherit },
            { "cacheable", Cacheable },
            { "cacheDurationSeconds", CacheDurationSeconds }
        };
    }

    /// <summary>
    /// Configuration for reusable authorization policies
    /// </summary>
    public record AuthzPolicyConfig(
        string Name,
        AuthzPolicyType Type = null!,
        string Description = "",
        string Rule = "",
        List<string> Attributes = null!,
        bool Cacheable = true,
        int CacheDurationSeconds = 300,
        bool Recursive = false,
        string Operations = "",
        bool AuditLogging = false,
        string ErrorMessage = ""
    )
    {
        public AuthzPolicyConfig(string name) : this(
            name,
            new AuthzPolicyType.Custom(),
            "",
            "",
            new List<string>(),
            true,
            300,
            false,
            "",
            false,
            ""
        ) { }

        public Dictionary<string, object> ToDict() => new()
        {
            { "name", Name },
            { "type", Type.Value },
            { "description", Description },
            { "rule", Rule },
            { "attributes", Attributes },
            { "cacheable", Cacheable },
            { "cacheDurationSeconds", CacheDurationSeconds },
            { "recursive", Recursive },
            { "operations", Operations },
            { "auditLogging", AuditLogging },
            { "errorMessage", ErrorMessage }
        };
    }

    /// <summary>
    /// Fluent builder for custom authorization rules
    /// </summary>
    public class AuthorizeBuilder
    {
        private string _rule = "";
        private string _policy = "";
        private string _description = "";
        private string _errorMessage = "";
        private bool _recursive = false;
        private string _operations = "";
        private bool _cacheable = true;
        private int _cacheDurationSeconds = 300;

        public AuthorizeBuilder Rule(string rule)
        {
            _rule = rule;
            return this;
        }

        public AuthorizeBuilder Policy(string policy)
        {
            _policy = policy;
            return this;
        }

        public AuthorizeBuilder Description(string description)
        {
            _description = description;
            return this;
        }

        public AuthorizeBuilder ErrorMessage(string errorMessage)
        {
            _errorMessage = errorMessage;
            return this;
        }

        public AuthorizeBuilder Recursive(bool recursive)
        {
            _recursive = recursive;
            return this;
        }

        public AuthorizeBuilder Operations(string operations)
        {
            _operations = operations;
            return this;
        }

        public AuthorizeBuilder Cacheable(bool cacheable)
        {
            _cacheable = cacheable;
            return this;
        }

        public AuthorizeBuilder CacheDurationSeconds(int duration)
        {
            _cacheDurationSeconds = duration;
            return this;
        }

        public AuthorizeConfig Build() =>
            new(
                _rule,
                _policy,
                _description,
                _errorMessage,
                _recursive,
                _operations,
                _cacheable,
                _cacheDurationSeconds
            );
    }

    /// <summary>
    /// Fluent builder for role-based access control
    /// </summary>
    public class RoleRequiredBuilder
    {
        private List<string> _roles = new();
        private RoleMatchStrategy _strategy = new RoleMatchStrategy.Any();
        private bool _hierarchy = false;
        private string _description = "";
        private string _errorMessage = "";
        private string _operations = "";
        private bool _inherit = false;
        private bool _cacheable = true;
        private int _cacheDurationSeconds = 300;

        public RoleRequiredBuilder Roles(params string[] roles)
        {
            _roles = roles.ToList();
            return this;
        }

        public RoleRequiredBuilder RolesArray(List<string> roles)
        {
            _roles = roles;
            return this;
        }

        public RoleRequiredBuilder Strategy(RoleMatchStrategy strategy)
        {
            _strategy = strategy;
            return this;
        }

        public RoleRequiredBuilder Hierarchy(bool hierarchy)
        {
            _hierarchy = hierarchy;
            return this;
        }

        public RoleRequiredBuilder Description(string description)
        {
            _description = description;
            return this;
        }

        public RoleRequiredBuilder ErrorMessage(string errorMessage)
        {
            _errorMessage = errorMessage;
            return this;
        }

        public RoleRequiredBuilder Operations(string operations)
        {
            _operations = operations;
            return this;
        }

        public RoleRequiredBuilder Inherit(bool inherit)
        {
            _inherit = inherit;
            return this;
        }

        public RoleRequiredBuilder Cacheable(bool cacheable)
        {
            _cacheable = cacheable;
            return this;
        }

        public RoleRequiredBuilder CacheDurationSeconds(int duration)
        {
            _cacheDurationSeconds = duration;
            return this;
        }

        public RoleRequiredConfig Build() =>
            new(
                _roles,
                _strategy,
                _hierarchy,
                _description,
                _errorMessage,
                _operations,
                _inherit,
                _cacheable,
                _cacheDurationSeconds
            );
    }

    /// <summary>
    /// Fluent builder for authorization policies
    /// </summary>
    public class AuthzPolicyBuilder
    {
        private readonly string _name;
        private AuthzPolicyType _type = new AuthzPolicyType.Custom();
        private string _description = "";
        private string _rule = "";
        private List<string> _attributes = new();
        private bool _cacheable = true;
        private int _cacheDurationSeconds = 300;
        private bool _recursive = false;
        private string _operations = "";
        private bool _auditLogging = false;
        private string _errorMessage = "";

        public AuthzPolicyBuilder(string name)
        {
            _name = name;
        }

        public AuthzPolicyBuilder Type(AuthzPolicyType type)
        {
            _type = type;
            return this;
        }

        public AuthzPolicyBuilder Description(string description)
        {
            _description = description;
            return this;
        }

        public AuthzPolicyBuilder Rule(string rule)
        {
            _rule = rule;
            return this;
        }

        public AuthzPolicyBuilder Attributes(params string[] attributes)
        {
            _attributes = attributes.ToList();
            return this;
        }

        public AuthzPolicyBuilder AttributesArray(List<string> attributes)
        {
            _attributes = attributes;
            return this;
        }

        public AuthzPolicyBuilder Cacheable(bool cacheable)
        {
            _cacheable = cacheable;
            return this;
        }

        public AuthzPolicyBuilder CacheDurationSeconds(int duration)
        {
            _cacheDurationSeconds = duration;
            return this;
        }

        public AuthzPolicyBuilder Recursive(bool recursive)
        {
            _recursive = recursive;
            return this;
        }

        public AuthzPolicyBuilder Operations(string operations)
        {
            _operations = operations;
            return this;
        }

        public AuthzPolicyBuilder AuditLogging(bool auditLogging)
        {
            _auditLogging = auditLogging;
            return this;
        }

        public AuthzPolicyBuilder ErrorMessage(string errorMessage)
        {
            _errorMessage = errorMessage;
            return this;
        }

        public AuthzPolicyConfig Build() =>
            new(
                _name,
                _type,
                _description,
                _rule,
                _attributes,
                _cacheable,
                _cacheDurationSeconds,
                _recursive,
                _operations,
                _auditLogging,
                _errorMessage
            );
    }

    /// <summary>
    /// Attribute for custom authorization rules
    /// </summary>
    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Property | AttributeTargets.Method)]
    public class AuthorizeAttribute : Attribute
    {
        public string Rule { get; set; } = "";
        public string Policy { get; set; } = "";
        public string Description { get; set; } = "";
        public string ErrorMessage { get; set; } = "";
        public bool Recursive { get; set; } = false;
        public string Operations { get; set; } = "";
        public bool Cacheable { get; set; } = true;
        public int CacheDurationSeconds { get; set; } = 300;
    }

    /// <summary>
    /// Attribute for role-based access control
    /// </summary>
    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Property | AttributeTargets.Method)]
    public class RoleRequiredAttribute : Attribute
    {
        public string[] Roles { get; set; } = Array.Empty<string>();
        public string Strategy { get; set; } = "any";
        public bool Hierarchy { get; set; } = false;
        public string Description { get; set; } = "";
        public string ErrorMessage { get; set; } = "";
        public string Operations { get; set; } = "";
        public bool Inherit { get; set; } = false;
        public bool Cacheable { get; set; } = true;
        public int CacheDurationSeconds { get; set; } = 300;
    }

    /// <summary>
    /// Attribute for authorization policies
    /// </summary>
    [AttributeUsage(AttributeTargets.Class | AttributeTargets.Property | AttributeTargets.Method)]
    public class AuthzPolicyAttribute : Attribute
    {
        public string Name { get; set; } = "";
        public string Type { get; set; } = "custom";
        public string Description { get; set; } = "";
        public string Rule { get; set; } = "";
        public string[] Attributes { get; set; } = Array.Empty<string>();
        public bool Cacheable { get; set; } = true;
        public int CacheDurationSeconds { get; set; } = 300;
        public bool Recursive { get; set; } = false;
        public string Operations { get; set; } = "";
        public bool AuditLogging { get; set; } = false;
        public string ErrorMessage { get; set; } = "";
    }
}
